tools/db_stress.cc

```diff
@@ -2457,28 +2465,158 @@ class StressTest {
+    for (int64_t rkey : rand_keys) {
+      std::string key_str = Key(rkey);
+      Slice key = key_str;
+
+      if (readoptionscopy.iterate_upper_bound != nullptr &&
+          thread->rand.OneIn(2)) {
+        // 1/2 chance, change the upper bound.
+        // It is possible that it is changed without first use, but there is no
+        // problem with that.
+        int64_t rand_upper_key = GenerateOneKey(thread, FLAGS_ops_per_thread);
+        upper_bound_str = Key(rand_upper_key);
+        upper_bound = Slice(upper_bound_str);
       }
-    }
 
-    if (s.ok()) {
-      thread->stats.AddIterations(1);
-    } else {
-      thread->stats.AddErrors(1);
+      // Set up an iterator and does the same without bounds and with total
+      // order seek and compare the results. This is to identify bugs related
+      // to bounds, prefix extractor or reseeking. Sometimes we are comparing
+      // iterators with the same set-up, and it doesn't hurt to check them
+      // to be equal.
+      ReadOptions cmp_ro;
+      cmp_ro.snapshot = snapshot;
+      cmp_ro.total_order_seek = true;
+      std::unique_ptr<Iterator> cmp_iter(
+          db_->NewIterator(readoptionscopy, cfh));
+      bool diverged = false;
+
+      iter->Seek(key);
+      cmp_iter->Seek(key);
+      VerifyIterator(thread, readoptionscopy, iter.get(), cmp_iter.get(), key,
+                     &diverged);
+
+      for (uint64_t i = 0; i < FLAGS_num_iterations && iter->Valid(); i++) {
+        if (thread->rand.OneIn(2)) {
+          iter->Next();
+          if (cmp_iter->Valid()) {
+            cmp_iter->Next();
+          }
+        } else {
+          iter->Prev();
+          if (cmp_iter->Valid()) {
+            cmp_iter->Prev();
+          }
+        }
+        VerifyIterator(thread, readoptionscopy, iter.get(), cmp_iter.get(), key,
+                       &diverged);
+      }
+
+      if (s.ok()) {
+        thread->stats.AddIterations(1);
+      } else {
+        thread->stats.AddErrors(1);
+        break;
+      }
     }
 
     db_->ReleaseSnapshot(snapshot);
 
     return s;
   }
 
+  // Compare the two iterator, iter and cmp_iter are in the same position,
+  // unless iter might be made invalidate or undefined because of
+  // upper or lower bounds, or prefix extractor.
+  // Will flag failure if the verification fails.
+  // diverged = true if the two iterator is already diverged.
+  // True if verification passed, false if not.
+  void VerifyIterator(ThreadState* thread, const ReadOptions& ro,
+                      Iterator* iter, Iterator* cmp_iter, const Slice& seek_key,
+                      bool* diverged) {
+    if (*diverged) {
+      return;
+    }
+
+    if (iter->Valid() && !cmp_iter->Valid()) {
+      fprintf(stderr,
+              "Control interator is invalid but iterator has value %s seek key "
+              "%s\n",
+              iter->key().ToString(true).c_str(),
+              seek_key.ToString(true).c_str());
+
+      *diverged = true;
```

Comment:
We set diverged to true but we do not set any error here!

Reply:
The error is set in line 2616.